middleware/timeout.go

```diff
@@ -0,0 +1,69 @@
+// +build go1.13
+
+package middleware
+
+import (
+	"context"
+	"github.com/labstack/echo/v4"
+	"net/http"
+	"time"
+)
+
+type (
+	// TimeoutConfig defines the config for Timeout middleware.
+	TimeoutConfig struct {
+		// Skipper defines a function to skip middleware.
+		Skipper Skipper
+		// Timeout configures a timeout for the middleware, defaults to 0 for no timeout
+		Timeout time.Duration
+	}
+)
+
+var (
+	// DefaultTimeoutConfig is the default Timeout middleware config.
+	DefaultTimeoutConfig = TimeoutConfig{
+		Skipper: DefaultSkipper,
+		Timeout: 0,
+	}
+)
+
+// Timeout returns a middleware which recovers from panics anywhere in the chain
+// and handles the control to the centralized HTTPErrorHandler.
+func Timeout() echo.MiddlewareFunc {
+	return TimeoutWithConfig(DefaultTimeoutConfig)
+}
+
+// TimeoutWithConfig returns a Timeout middleware with config.
+// See: `Timeout()`.
+func TimeoutWithConfig(config TimeoutConfig) echo.MiddlewareFunc {
+	// Defaults
+	if config.Skipper == nil {
+		config.Skipper = DefaultTimeoutConfig.Skipper
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) || config.Timeout == 0 {
+				return next(c)
+			}
+
+			ctx, cancel := context.WithTimeout(c.Request().Context(), config.Timeout)
+			defer cancel()
+
+			// this does a deep clone of the context, wondering if there is a better way to do this?
+			c.SetRequest(c.Request().Clone(ctx))
+
+			done := make(chan error, 1)
+			go func() {
+				done <- next(c)
+			}()
+
+			select {
+			case <-ctx.Done():
+				return c.JSON(http.StatusGatewayTimeout, ctx.Err())
```

Comment:
This is assuming that user of this middleware is returning JSON responses, which couldn't be true. Please add a function to TimeoutConfig that will be called here as an error handler.

Reply:
Agreed, I will add an error handler, and if no error handler is defined it will return the error back and it can be handled by whichever middleware we have for the errors.